python/triton/language/standard.py

```diff
@@ -260,11 +260,24 @@ def _sum_combine(a, b):
 
 
 @core._tensor_member_fn
-@jit
-@core._add_reduction_docstr("sum")
-def sum(input, axis=None, keep_dims=False):
-    input = core._promote_bfloat16_to_float32(input)
-    return core.reduce(input, axis, _sum_combine, keep_dims=keep_dims)
+@core.builtin
+@core._add_reduction_docstr("sum", dtype_arg="dtype")
+def sum(input, axis=None, keep_dims=False, dtype=None, _builder=None, _generator=None):
+    # Pick a default dtype for the reduction if one was not specified.
+    if dtype is None:
+        # For integer bitwidths less than 32, pick int32 with the same sign to
+        # avoid overflow.
+        if input.dtype.is_int_signed():
+            dtype = core.int32 if input.dtype.int_bitwidth < 32 else None
+        elif input.dtype.is_int_unsigned():
+            dtype = core.uint32 if input.dtype.int_bitwidth < 32 else None
+        # For float bitwidths less than 32, pick float32 to avoid overflow or
+        # precision loss.
+        elif input.dtype.is_floating():
+            dtype = core.float32 if input.dtype.primitive_bitwidth < 32 else None
+
+    input = core._promote_bfloat16_to_float32(input, _builder=_builder)
+    return core.reduce(input, axis, _sum_combine, keep_dims=keep_dims, dtype=dtype, _builder=_builder, _generator=_generator)
```

Comment:
This promotion is redundant, plus I don't think we should change `reduce`.
```suggestion
    input = input.to(dtype, _builder=_builder)
    return core.reduce(input, axis, _sum_combine, keep_dims=keep_dims, _builder=_builder, _generator=_generator)
```

Reply:
makes sense